File: diary-server/src/main/java/com/diary/server/service/DiaryService.java
package com.diary.server.service;

import com.diary.server.domain.Diary;
import com.diary.server.domain.User;
import com.diary.server.repository.DiaryRepository;
import com.diary.server.repository.UserRepository;
import com.diary.server.web.dto.DiaryRequest;
import com.diary.server.web.dto.DiaryResponse;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

@Service
public class DiaryService {

    private final DiaryRepository diaryRepository;
    private final UserRepository userRepository;

    public DiaryService(DiaryRepository diaryRepository, UserRepository userRepository) {
        this.diaryRepository = diaryRepository;
        this.userRepository = userRepository;
    }

    @Transactional(readOnly = true)
    public List<DiaryResponse> listDiaries(Long userId, Instant since) {
        User user = requireUser(userId);
        return diaryRepository.findActiveByUserSince(user, since).stream()
                .map(DiaryResponse::from)
                .toList();
    }

    @Transactional(readOnly = true)
    public DiaryResponse getDiary(Long userId, LocalDate date) {
        User user = requireUser(userId);
        Diary diary = diaryRepository.findByUserAndDiaryDateAndDeletedAtIsNull(user, date)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Diary not found"));
        return DiaryResponse.from(diary);
    }

    @Transactional
    public DiaryResponse upsertDiary(Long userId, LocalDate date, DiaryRequest request) {
        User user = requireUser(userId);
        Diary diary = diaryRepository.findByUserAndDiaryDateAndDeletedAtIsNull(user, date)
                .orElseGet(() -> {
                    Diary created = new Diary();
                    created.setUser(user);
                    created.setDiaryDate(date);
                    return created;
                });

        diary.setContent(request.content() == null ? "" : request.content());
        diary.setLocation(request.location());
        diary.setLatitude(request.latitude());
        diary.setLongitude(request.longitude());
        diary.setDeletedAt(null);

        return DiaryResponse.from(diaryRepository.save(diary));
    }

    @Transactional
    public void deleteDiary(Long userId, LocalDate date) {
        User user = requireUser(userId);
        Diary diary = diaryRepository.findByUserAndDiaryDateAndDeletedAtIsNull(user, date)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Diary not found"));
        diary.setDeletedAt(Instant.now());
        diaryRepository.save(diary);
    }

    private User requireUser(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User not found"));
    }
}

File: diary-server/src/main/kotlin/com/diary/server/service/DiaryService.kt
package com.diary.server.service

import com.diary.server.domain.Diary
import com.diary.server.domain.User
import com.diary.server.repository.DiaryRepository
import com.diary.server.repository.UserRepository
import com.diary.server.web.dto.DiaryRequest
import com.diary.server.web.dto.DiaryResponse
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException
import java.time.Instant
import java.time.LocalDate

@Service
class DiaryService(
    private val diaryRepository: DiaryRepository,
    private val userRepository: UserRepository
) {
    @Transactional(readOnly = true)
    fun listDiaries(userId: Long, since: Instant?): List<DiaryResponse> {
        val user = requireUser(userId)
        return diaryRepository.findActiveByUserSince(user, since)
            .map { DiaryResponse.from(it) }
    }

    @Transactional(readOnly = true)
    fun getDiary(userId: Long, date: LocalDate): DiaryResponse {
        val user = requireUser(userId)
        val diary = diaryRepository.findByUserAndDiaryDateAndDeletedAtIsNull(user, date)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Diary not found")
        return DiaryResponse.from(diary)
    }

    @Transactional
    fun upsertDiary(userId: Long, date: LocalDate, request: DiaryRequest): DiaryResponse {
        val user = requireUser(userId)
        val diary = diaryRepository.findByUserAndDiaryDateAndDeletedAtIsNull(user, date)
            ?: Diary().apply {
                this.user = user
                diaryDate = date
            }

        diary.content = request.content ?: ""
        diary.location = request.location
        diary.latitude = request.latitude
        diary.longitude = request.longitude
        diary.deletedAt = null

        return DiaryResponse.from(diaryRepository.save(diary))
    }

    @Transactional
    fun deleteDiary(userId: Long, date: LocalDate) {
        val user = requireUser(userId)
        val diary = diaryRepository.findByUserAndDiaryDateAndDeletedAtIsNull(user, date)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Diary not found")
        diary.deletedAt = Instant.now()
        diaryRepository.save(diary)
    }

    private fun requireUser(userId: Long): User =
        userRepository.findById(userId).orElseThrow {
            ResponseStatusException(HttpStatus.UNAUTHORIZED, "User not found")
        }
}
